### PR TITLE
Bugfix/prevent overflow with mtime

### PR DESCRIPTION
**Pull request by mgallien**

### CODE REVIEW

**Review by claucambra — APPROVED** [review]

File: src/libsync/discovery.cpp [hunk 1]
```diff
@@ -539,18 +539,18 @@ void ProcessDirectoryJob::processFileAnalyzeRemoteInfo(
             } else {
                 item->_instruction = CSYNC_INSTRUCTION_SYNC;
             }
-        } else if (dbEntry._modtime <= 0 && serverEntry.modtime > 0) {
+        } else if ((dbEntry._modtime <= 0 && serverEntry.modtime > 0) || dbEntry._modtime != serverEntry.modtime) {
```
Comment (allexzander) [inline src/libsync/discovery.cpp:542]: @mgallien Doesn't `dbEntry._modtime != serverEntry.modtime` also cover the case when `(dbEntry._modtime <= 0 && serverEntry.modtime > 0)`? If this is the case, can we just go with `if(dbEntry._modtime != serverEntry.modtime)`?

**Comment by PhilippSchlesinger** [comment]
Is this PR intended to fix #4188?

File: src/libsync/discovery.cpp [hunk 2]
```diff
@@ -539,18 +539,18 @@ void ProcessDirectoryJob::processFileAnalyzeRemoteInfo(
...
-                // Deleted locally, changed on server
-                item->_instruction = CSYNC_INSTRUCTION_NEW;
-            } else {
-                item->_instruction = CSYNC_INSTRUCTION_SYNC;
+            item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+            if (dbEntry._modtime != serverEntry.modtime && dbEntry._etag == serverEntry.etag) {
+                qCWarning(lcDisco()) << "server modification time has been modified without changing etag: force full discovery";
+                _forceRemoteDiscovery = true;
+            }
+            if (dbEntry._modtime != serverEntry.modtime && dbEntry._etag == serverEntry.etag) {
```
Comment (allexzander) [inline src/libsync/discovery.cpp:551]: @mgallien Why do you need to repeat the same logic here?

File: src/libsync/discoveryphase.cpp [hunk 3]
```diff
@@ -396,7 +396,11 @@ static void propertyMapToRemoteInfo(const QMap<QString, QString> &map, RemoteInf
         } else if (property == QLatin1String("getlastmodified")) {
             const auto date = QDateTime::fromString(value, Qt::RFC2822Date);
             Q_ASSERT(date.isValid());
-            result.modtime = date.toTime_t();
+            if (date.toSecsSinceEpoch() > 0) {
+                result.modtime = date.toSecsSinceEpoch();
+            } else {
+                result.modtime = 0;
```
Comment (allexzander) [inline src/libsync/discoveryphase.cpp:402]: @mgallien You can get rid of `else` block if you move `result.modtime = 0;` before `if (date.toSecsSinceEpoch() > 0) {`, bu, just a nitpik

File: src/libsync/propagatedownload.cpp [hunk 4]
```diff
@@ -563,7 +563,6 @@ void PropagateDownloadFile::startAfterIsEncryptedIsChecked()
         return checksum_header.startsWith("SHA")
             || checksum_header.startsWith("MD5:");
     };
-    Q_ASSERT(_item->_modtime > 0);
```
Comment (allexzander) [inline src/libsync/propagatedownload.cpp:566]: @mgallien Why remove this?

File: test/syncenginetestutils.cpp [hunk 5]
```diff
@@ -188,6 +188,13 @@ void FileInfo::setModTime(const QString &relativePath, const QDateTime &modTime)
     file->lastModified = modTime;
 }
 
+void FileInfo::setModTimeKeepEtag(const QString &relativePath, const QDateTime &modTime)
+{
+    FileInfo *file = find(relativePath);
+    Q_ASSERT(file);
+    file->lastModified = modTime;
```
Comment (allexzander) [inline test/syncenginetestutils.cpp:195]: @mgallien don't you want to check the pointer before resolution? Such tat we don't have a crash in runtime. And maybe a log if it is `null`.

File: src/libsync/discovery.cpp (hunk 1 above)
Comment (mgallien) [reply src/libsync/discovery.cpp:542]: yes

File: src/libsync/discovery.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/discovery.cpp:551]: I will check

File: src/libsync/discoveryphase.cpp (hunk 3 above)
Comment (mgallien) [reply src/libsync/discoveryphase.cpp:402]: yes good idea

File: src/libsync/propagatedownload.cpp (hunk 4 above)
Comment (mgallien) [reply src/libsync/propagatedownload.cpp:566]: this makes it impossible to sync against a server with files that have a `modtime` equal to `0`
in that case the assert will fire and crash the application
the code is indeed able to deal with that situation now so the assert is not needed

File: test/syncenginetestutils.cpp (hunk 5 above)
Comment (mgallien) [reply test/syncenginetestutils.cpp:195]: that was what I wanted to do with `Q_ASSERT(file);` as we are in the automated test code and I expect people to build it in debug mode

**Comment by mgallien** [comment]
> Is this PR intended to fix #4188?

not sure because it depends on the exact state of your client sync folder
@PhilippSchlesinger maybe you could share your logs and I can have a quick look

File: src/libsync/discovery.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/discovery.cpp]: thanks that part got duplicated during rebase
thanks for catching it

**Comment by PhilippSchlesinger** [comment]
@mgallien: I'd be happy to share my logs. As the logs contain confidential information, how can I provide the logs privately?

File: src/libsync/discovery.cpp [hunk 6]
```diff
@@ -539,18 +539,14 @@ void ProcessDirectoryJob::processFileAnalyzeRemoteInfo(
...
-            if (serverEntry.isDirectory) {
-                ENFORCE(dbEntry.isDirectory());
-                item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
-            } else if (!localEntry.isValid() && _queryLocal != ParentNotChanged) {
-                // Deleted locally, changed on server
-                item->_instruction = CSYNC_INSTRUCTION_NEW;
-            } else {
-                item->_instruction = CSYNC_INSTRUCTION_SYNC;
+            item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+            if (dbEntry._modtime != serverEntry.modtime && dbEntry._etag == serverEntry.etag) {
```
Comment (allexzander) [inline src/libsync/discovery.cpp:547]: @mgallien Why do you need this condition? The parent condition `else if` already has it

Comment (mgallien) [reply src/libsync/discovery.cpp]: thanks
I removed this code that was not really working well and not needed

**Comment by PhilippSchlesinger** [comment]
> @mgallien: I'd be happy to share my logs. As the logs contain confidential information, how can I provide the logs privately?

@mgallien: could you please give some advice how and which part of the log zip file to provide?
thx

**Review by claucambra — APPROVED** [review]

**Review by allexzander — APPROVED** [review]

**Pull request merged**